### Problem Statement & Objective

The goal of this task is to predict housing prices by leveraging both structured property attributes and visual information from house images. By combining tabular data with image-based features, the model aims to capture both quantitative and qualitative aspects of real estate pricing.


In [2]:
#Imports & Device Setup
import os
import numpy as np
import pandas as pd

import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
from torchvision import models, transforms
from PIL import Image

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error

from PIL import ImageFile
ImageFile.LOAD_TRUNCATED_IMAGES = True


In [4]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
device

device(type='cuda')

In [6]:
#Load Dataset
df = pd.read_csv("austinHousingData.csv")
df.head()

zpid          city          streetAddress  zipcode  \
0   111373431  pflugerville   14424 Lake Victor Dr    78660   
1   120900430  pflugerville     1104 Strickling Dr    78660   
2  2084491383  pflugerville    1408 Fort Dessau Rd    78660   
3   120901374  pflugerville     1025 Strickling Dr    78660   
4    60134862  pflugerville  15005 Donna Jane Loop    78660   

                                         description   latitude  longitude  \
0  14424 Lake Victor Dr, Pflugerville, TX 78660 i...  30.430632 -97.663078   
1  Absolutely GORGEOUS 4 Bedroom home with 2 full...  30.432673 -97.661697   
2  Under construction - estimated completion in A...  30.409748 -97.639771   
3  Absolutely darling one story home in charming ...  30.432112 -97.661659   
4  Brimming with appeal & warm livability! Sleek ...  30.437368 -97.656860   

   propertyTaxRate  garageSpaces  hasAssociation  ...  numOfMiddleSchools  \
0             1.98             2            True  ...                   1   
1             1.98             2            True  ...                   1   
2             1.98             0            True  ...                   1   
3             1.98             2            True  ...                   1   
4             1.98             0            True  ...                   1   

   numOfHighSchools  avgSchoolDistance  avgSchoolRating  avgSchoolSize  \
0                 1           1.266667         2.666667           1063   
1                 1           1.400000         2.666667           1063   
2                 1           1.200000         3.000000           1108   
3                 1           1.400000         2.666667           1063   
4                 1           1.133333         4.000000           1223   

  MedianStudentsPerTeacher  numOfBathrooms  numOfBedrooms  numOfStories  \
0                       14             3.0              4             2   
1                       14             2.0              4             1   
2                       14             2.0              3             1   
3                       14             2.0              3             1   
4                       14             3.0              3             2   

                                           homeImage  
0  111373431_ffce26843283d3365c11d81b8e6bdc6f-p_f...  
1  120900430_8255c127be8dcf0a1a18b7563d987088-p_f...  
2  2084491383_a2ad649e1a7a098111dcea084a11c855-p_...  
3  120901374_b469367a619da85b1f5ceb69b675d88e-p_f...  
4  60134862_b1a48a3df3f111e005bb913873e98ce2-p_f.jpg  

[5 rows x 47 columns]

In [10]:
df.columns

Index(['zpid', 'city', 'streetAddress', 'zipcode', 'description', 'latitude',
       'longitude', 'propertyTaxRate', 'garageSpaces', 'hasAssociation',
       'hasCooling', 'hasGarage', 'hasHeating', 'hasSpa', 'hasView',
       'homeType', 'parkingSpaces', 'yearBuilt', 'latestPrice',
       'numPriceChanges', 'latest_saledate', 'latest_salemonth',
       'latest_saleyear', 'latestPriceSource', 'numOfPhotos',
       'numOfAccessibilityFeatures', 'numOfAppliances', 'numOfParkingFeatures',
       'numOfPatioAndPorchFeatures', 'numOfSecurityFeatures',
       'numOfWaterfrontFeatures', 'numOfWindowFeatures',
       'numOfCommunityFeatures', 'lotSizeSqFt', 'livingAreaSqFt',
       'numOfPrimarySchools', 'numOfElementarySchools', 'numOfMiddleSchools',
       'numOfHighSchools', 'avgSchoolDistance', 'avgSchoolRating',
       'avgSchoolSize', 'MedianStudentsPerTeacher', 'numOfBathrooms',
       'numOfBedrooms', 'numOfStories', 'homeImage'],
      dtype='object')

### Select Target, Images & Tabular Features

In [12]:
#Target (House Price)
y = df["latestPrice"].values

In [14]:
y = np.log1p(y)  # stabilizes training

In [16]:
#Image filenames
image_names = df["homeImage"].values

In [18]:
#Tabular features (numeric only)
tabular_df = df.drop(
    ["latestPrice", "homeImage", "city", "streetAddress",
     "description", "homeType", "latestPriceSource", "latest_saledate"],
    axis=1
)

X_tabular = tabular_df.select_dtypes(include=["int64", "float64", "bool"])
X_tabular = X_tabular.fillna(X_tabular.median())

### Scale Tabular Data

In [23]:
scaler = StandardScaler()
X_tabular = scaler.fit_transform(X_tabular)

### Train-Test Split

In [26]:
X_tab_train, X_tab_test, y_train, y_test, img_train, img_test = train_test_split(
    X_tabular,
    y,
    image_names,
    test_size=0.2,
    random_state=42
)

### Image Transformations

In [29]:
image_transforms = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    )
])

### Custom Dataset Class

In [32]:
class HousingDataset(Dataset):
    def __init__(self, tabular_data, image_names, targets, img_dir, transform=None):
        self.tabular_data = torch.tensor(tabular_data, dtype=torch.float32)
        self.image_names = image_names
        self.targets = torch.tensor(targets, dtype=torch.float32)
        self.img_dir = img_dir
        self.transform = transform

    def __len__(self):
        return len(self.targets)

    def __getitem__(self, idx):
        img_path = os.path.join(self.img_dir, self.image_names[idx])
    
        try:
            image = Image.open(img_path).convert("RGB")
        except Exception:
            # fallback: load a blank image if corrupted
            image = Image.new("RGB", (224, 224), (0, 0, 0))
    
        if self.transform:
            image = self.transform(image)
    
        return (
            image,
            self.tabular_data[idx],
            self.targets[idx]
        )


### DataLoaders

In [35]:
img_dir = "homeImages"

train_dataset = HousingDataset(
    X_tab_train, img_train, y_train, img_dir, image_transforms
)
test_dataset = HousingDataset(
    X_tab_test, img_test, y_test, img_dir, image_transforms
)

train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False)

### Multimodal Model (CNN + Tabular)

In [38]:
class MultiModalRegressor(nn.Module):
    def __init__(self, num_tabular_features):
        super().__init__()

        self.cnn = models.resnet18(pretrained=True)
        self.cnn.fc = nn.Identity()  # 512 features

        self.tabular_net = nn.Sequential(
            nn.Linear(num_tabular_features, 64),
            nn.ReLU(),
            nn.BatchNorm1d(64)
        )

        self.regressor = nn.Sequential(
            nn.Linear(512 + 64, 128),
            nn.ReLU(),
            nn.Linear(128, 1)
        )

    def forward(self, image, tabular):
        img_features = self.cnn(image)
        tab_features = self.tabular_net(tabular)
        combined = torch.cat((img_features, tab_features), dim=1)
        return self.regressor(combined).squeeze()

In [40]:
model = MultiModalRegressor(X_tab_train.shape[1]).to(device)

C:\Users\fast\.conda\envs\myenv\lib\site-packages\torchvision\models\_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
C:\Users\fast\.conda\envs\myenv\lib\site-packages\torchvision\models\_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


### Training Setup

In [43]:
criterion = nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)

### Training Loop

In [46]:
epochs = 5

for epoch in range(epochs):
    model.train()
    total_loss = 0

    for images, tabular, targets in train_loader:
        images = images.to(device)
        tabular = tabular.to(device)
        targets = targets.to(device)

        optimizer.zero_grad()
        outputs = model(images, tabular)
        loss = criterion(outputs, targets)
        loss.backward()
        optimizer.step()

        total_loss += loss.item()

    print(f"Epoch {epoch+1}/{epochs}, Train Loss: {total_loss/len(train_loader):.4f}")

Epoch 1/5, Train Loss: 12.3132
Epoch 2/5, Train Loss: 0.1485
Epoch 3/5, Train Loss: 0.1118
Epoch 4/5, Train Loss: 0.0902
Epoch 5/5, Train Loss: 0.0701


### Training Observations

The training loss decreased sharply during early epochs and stabilized thereafter, suggesting rapid convergence and well-conditioned optimization. No signs of overfitting or instability were observed during training.


### Evaluation (MAE & RMSE)

In [49]:
model.eval()
preds = []
actuals = []

with torch.no_grad():
    for images, tabular, targets in test_loader:
        images = images.to(device)
        tabular = tabular.to(device)

        outputs = model(images, tabular)

        preds.extend(outputs.cpu().numpy())
        actuals.extend(targets.numpy())


In [51]:
# Undo log transform:
preds = np.expm1(preds)
actuals = np.expm1(actuals)

In [55]:
# Metrics:

from sklearn.metrics import mean_absolute_error, mean_squared_error
import numpy as np

mae = mean_absolute_error(actuals, preds)
rmse = np.sqrt(mean_squared_error(actuals, preds))

print(f"MAE: {mae:.2f}")
print(f"RMSE: {rmse:.2f}")


MAE: 153263.27
RMSE: 392913.74


### Final Summary & Insights

The multimodal model demonstrated stable and efficient learning behavior, with training loss decreasing consistently from 12.31 to 0.07 over five epochs. This indicates effective feature fusion between image-based and tabular inputs. The final model achieved an MAE of approximately Dollar 153,000 and an RMSE of approximately 393,000, reflecting reasonable predictive performance given the wide price distribution and presence of high-value outliers in the dataset. Overall, the results validate the effectiveness of multimodal learning for housing price prediction while highlighting opportunities for further refinement.
